### PR TITLE
Upstream of various fixes and features (details below)

### PR DESCRIPTION
**Pull request by glennw**
Add missing get_api stubs.
Add hidpi_factor() support for retina displays.
Fix Resize and MouseMoved events to handle retina displays.
Fix inverted y position for MouseMoved events on mac.
Fix initial painting on retina display.

### CODE REVIEW

File: src/android/mod.rs [hunk 1]
```diff
@@ -63,6 +63,10 @@ impl HeadlessContext {
     pub fn get_proc_address(&self, _addr: &str) -> *const () {
         unimplemented!()
     }
+
+    pub fn get_api(&self) -> ::Api {
+        ::Api::OpenGl
```
Comment (ozkriff) [inline src/android/mod.rs]: ::Api::OpenGlEs?

Comment (glennw) [inline src/android/mod.rs]: Ooops, fixed!

**Comment by tomaka** [comment]
Does that work transparently for the user?
Does the user need to use the hdipi factor on the mouse positions, dimensions, etc. or is that done automatically?

**Comment by glennw** [comment]
Mouse positions, resize events are in device pixels. Generally this is what you want and it all works transparently. Occasionally you might need to use the hidpi_factor() if you want to convert the pixel values to the virtual window size units.

**Pull request merged**